File: apps/desktop/electron/scm.ts
import { spawn } from "node:child_process";
import path from "node:path";
import type {
  GitStatusResult,
  GitFileStatus,
  GitDiffResult,
  GitStageArgs,
  GitCommitArgs,
  GitCommitResult,
  GitRunCommand,
  GitRunResult,
} from "@idex/types";

/**
 * Source-control bridge.
 *
 * Shells out to `git` rather than embedding `simple-git` because (a)
 * git is universally installed, (b) we get exact CLI semantics, (c) we
 * avoid bundling another 3MB of node-modules. Every call uses
 * `--no-pager` to stop git from spawning `less` and hanging us.
 *
 * Path safety: every path argument coming from the renderer is
 * resolved against the workspace root and refused if it escapes.
 */

interface RunOpts {
  cwd: string;
  /** Args after `git`. Do NOT include `git` itself. */
  args: string[];
  /** stdin to pipe in (used for commit message). */
  stdin?: string;
  /** Cap how long the command can run. */
  timeoutMs?: number;
}

interface RunResult {
  code: number;
  stdout: string;
  stderr: string;
}

function runGit(opts: RunOpts): Promise<RunResult> {
  return new Promise((resolve) => {
    const child = spawn("git", ["--no-pager", ...opts.args], { cwd: opts.cwd });
    let stdout = "";
    let stderr = "";
    let killed = false;
    const t = opts.timeoutMs
      ? setTimeout(() => {
          killed = true;
          child.kill("SIGTERM");
        }, opts.timeoutMs)
      : null;
    child.stdout.on("data", (d: Buffer) => (stdout += d.toString("utf8")));
    child.stderr.on("data", (d: Buffer) => (stderr += d.toString("utf8")));
    child.on("error", () => {
      if (t) clearTimeout(t);
      resolve({ code: -1, stdout, stderr: stderr || "git not installed" });
    });
    child.on("close", (code) => {
      if (t) clearTimeout(t);
      if (killed) {
        resolve({ code: -1, stdout, stderr: stderr + "\n(git command timed out)" });
        return;
      }
      resolve({ code: code ?? -1, stdout, stderr });
    });
    if (opts.stdin && child.stdin) {
      child.stdin.write(opts.stdin);
      child.stdin.end();
    }
  });
}

/**
 * Resolve a workspace-relative path safely. Refuses absolute paths and
 * any traversal that escapes the root.
 */
function safeRelative(rootPath: string, rel: string): string | null {
  if (path.isAbsolute(rel)) return null;
  const abs = path.resolve(rootPath, rel);
  if (!abs.startsWith(rootPath + path.sep) && abs !== rootPath) return null;
  return path.relative(rootPath, abs);
}

export async function gitStatus(rootPath: string): Promise<GitStatusResult> {
  if (!rootPath) {
    return { ok: false, branch: null, ahead: 0, behind: 0, files: [], error: "No workspace open" };
  }

  // Verify we're inside a git repo.
  const inside = await runGit({
    cwd: rootPath,
    args: ["rev-parse", "--is-inside-work-tree"],
    timeoutMs: 3000,
  });
  if (inside.code !== 0) {
    return {
      ok: false,
      branch: null,
      ahead: 0,
      behind: 0,
      files: [],
      error: "Not a git repository",
    };
  }

  // Branch + ahead/behind in one porcelain v2 call.
  const status = await runGit({
    cwd: rootPath,
    args: ["status", "--porcelain=v2", "--branch", "--untracked-files=all"],
    timeoutMs: 5000,
  });
  if (status.code !== 0) {
    return {
      ok: false,
      branch: null,
      ahead: 0,
      behind: 0,
      files: [],
      error: status.stderr.trim() || "git status failed",
    };
  }

  return parseStatus(status.stdout);
}

/**
 * Parse git's porcelain v2 output. Each line:
 *   `# branch.head <name>`
 *   `# branch.ab +<ahead> -<behind>`
 *   `1 XY ... <path>`               — tracked changed file
 *   `2 XY ... <orig> <path>`        — renamed/copied file
 *   `? <path>`                      — untracked file
 */
function parseStatus(output: string): GitStatusResult {
  let branch: string | null = null;
  let ahead = 0;
  let behind = 0;
  const files: GitFileStatus[] = [];

  for (const line of output.split("\n")) {
    if (!line) continue;
    if (line.startsWith("# branch.head ")) {
      const name = line.slice("# branch.head ".length).trim();
      branch = name === "(detached)" ? null : name;
    } else if (line.startsWith("# branch.ab ")) {
      const m = line.match(/\+(\d+)\s+-(\d+)/);
      if (m) {
        ahead = Number(m[1]);
        behind = Number(m[2]);
      }
    } else if (line.startsWith("1 ") || line.startsWith("2 ")) {
      // Format: "1 XY sub-mode HEAD-mode worktree-mode HEAD-sha worktree-sha path"
      const parts = line.split(" ");
      const xy = parts[1] ?? "..";
      const x = xy[0] ?? ".";
      const y = xy[1] ?? ".";
      const filePath = line.startsWith("2 ")
        ? (parts[parts.length - 2] ?? "")
        : (parts[parts.length - 1] ?? "");
      if (!filePath) continue;
      files.push({
        path: filePath,
        index: x,
        workingTree: y,
        staged: x !== "." && x !== "?",
      });
    } else if (line.startsWith("? ")) {
      files.push({
        path: line.slice(2),
        index: "?",
        workingTree: "?",
        staged: false,
      });
    }
  }

  return { ok: true, branch, ahead, behind, files };
}

export async function gitDiff(
  rootPath: string,
  rel: string,
  staged = false,
): Promise<GitDiffResult> {
  const safe = safeRelative(rootPath, rel);
  if (!safe) return { ok: false, diff: "", error: "Path escapes workspace" };

  const args = ["diff", "-U3", "--no-color"];
  if (staged) args.push("--cached");
  args.push("--", safe);

  const r = await runGit({ cwd: rootPath, args, timeoutMs: 10000 });
  if (r.code !== 0 && r.code !== 1) {
    return { ok: false, diff: "", error: r.stderr.trim() || "git diff failed" };
  }
  return { ok: true, diff: r.stdout };
}

export async function gitStage(
  rootPath: string,
  args: GitStageArgs,
): Promise<{ ok: boolean; error?: string }> {
  const paths: string[] = [];
  for (const p of args.paths) {
    const safe = safeRelative(rootPath, p);
    if (!safe) return { ok: false, error: `Path escapes workspace: ${p}` };
    paths.push(safe);
  }
  const r = await runGit({
    cwd: rootPath,
    args: [args.stage ? "add" : "reset", "--", ...paths],
    timeoutMs: 10000,
  });
  return r.code === 0 ? { ok: true } : { ok: false, error: r.stderr.trim() };
}

export async function gitCommit(
  rootPath: string,
  args: GitCommitArgs,
): Promise<GitCommitResult> {
  const message = (args.message ?? "").trim();
  if (!message) return { ok: false, error: "Commit message is empty" };

  if (args.stageAll) {
    const stage = await runGit({ cwd: rootPath, args: ["add", "-A"], timeoutMs: 10000 });
    if (stage.code !== 0) {
      return { ok: false, error: stage.stderr.trim() || "git add -A failed" };
    }
  }

  const r = await runGit({
    cwd: rootPath,
    args: ["commit", "-F", "-"],
    stdin: message,
    timeoutMs: 15000,
  });
  if (r.code !== 0) {
    return { ok: false, error: r.stderr.trim() || r.stdout.trim() };
  }

  // Pull the new sha.
  const sha = await runGit({ cwd: rootPath, args: ["rev-parse", "HEAD"], timeoutMs: 3000 });
  return {
    ok: true,
    sha: sha.code === 0 ? sha.stdout.trim() : undefined,
  };
}

const ALLOWED_RUN: GitRunCommand[] = ["pull", "push", "fetch"];

export async function gitRun(
  rootPath: string,
  cmd: GitRunCommand,
): Promise<GitRunResult> {
  if (!ALLOWED_RUN.includes(cmd)) {
    return { ok: false, error: `Command not allowed: ${cmd}` };
  }
  const r = await runGit({
    cwd: rootPath,
    args: [cmd],
    timeoutMs: 60_000,
  });
  return {
    ok: r.code === 0,
    stdout: r.stdout,
    stderr: r.stderr,
    error: r.code === 0 ? undefined : r.stderr.trim() || `git ${cmd} failed`,
  };
}

File: apps/desktop/src-tauri/src/scm.rs
//! Source control bridge — Rust port of `electron/scm.ts`.
//!
//! Shells out to `git --no-pager` so we never spawn `less`. Same IPC
//! shape as the Electron handlers: status / diff / stage / commit / run.
//! Path-safety: every workspace-relative path argument is resolved
//! against the root and refused if it escapes.

use serde::{Deserialize, Serialize};
use std::io::Write;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::time::Duration;

/* ────────────────────────────────────────── *
 * IPC payload types                          *
 * ────────────────────────────────────────── */

#[derive(Serialize, Clone)]
pub struct GitFileStatus {
    pub path: String,
    pub index: String,
    #[serde(rename = "workingTree")]
    pub working_tree: String,
    pub staged: bool,
}

#[derive(Serialize)]
pub struct GitStatusResult {
    pub ok: bool,
    pub branch: Option<String>,
    pub ahead: u32,
    pub behind: u32,
    pub files: Vec<GitFileStatus>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Serialize)]
pub struct GitDiffResult {
    pub ok: bool,
    pub diff: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Deserialize)]
pub struct StageArgs {
    pub paths: Vec<String>,
    pub stage: bool,
}

#[derive(Deserialize)]
pub struct CommitArgs {
    pub message: String,
    #[serde(default, rename = "stageAll")]
    pub stage_all: bool,
}

#[derive(Serialize)]
pub struct CommitResult {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sha: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Serialize)]
pub struct StageResult {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Serialize)]
pub struct RunResult {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stdout: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stderr: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/* ────────────────────────────────────────── *
 * git runner                                 *
 * ────────────────────────────────────────── */

struct GitRun {
    code: i32,
    stdout: String,
    stderr: String,
}

fn run_git(cwd: &str, args: &[&str], stdin: Option<&str>, timeout: Duration) -> GitRun {
    let mut cmd = Command::new("git");
    cmd.arg("--no-pager");
    for a in args {
        cmd.arg(a);
    }
    cmd.current_dir(cwd);
    if stdin.is_some() {
        cmd.stdin(Stdio::piped());
    }
    cmd.stdout(Stdio::piped());
    cmd.stderr(Stdio::piped());

    let mut child = match cmd.spawn() {
        Ok(c) => c,
        Err(e) => {
            return GitRun {
                code: -1,
                stdout: String::new(),
                stderr: if e.kind() == std::io::ErrorKind::NotFound {
                    "git not installed".to_string()
                } else {
                    e.to_string()
                },
            };
        }
    };

    if let Some(input) = stdin {
        if let Some(mut sin) = child.stdin.take() {
            let _ = sin.write_all(input.as_bytes());
        }
    }

    // Naive timeout via wait + sleep loop. portable-pty's child::wait
    // is blocking; we approximate a deadline by polling try_wait every
    // 50ms and SIGTERM-ing once we exceed the budget.
    let deadline = std::time::Instant::now() + timeout;
    loop {
        match child.try_wait() {
            Ok(Some(_)) => break,
            Ok(None) => {
                if std::time::Instant::now() >= deadline {
                    let _ = child.kill();
                    let out = child.wait_with_output().ok();
                    return GitRun {
                        code: -1,
                        stdout: String::new(),
                        stderr: out
                            .map(|o| String::from_utf8_lossy(&o.stderr).to_string())
                            .unwrap_or_else(|| "git timed out".to_string()),
                    };
                }
                std::thread::sleep(Duration::from_millis(50));
            }
            Err(e) => {
                return GitRun {
                    code: -1,
                    stdout: String::new(),
                    stderr: e.to_string(),
                };
            }
        }
    }

    let output = match child.wait_with_output() {
        Ok(o) => o,
        Err(e) => {
            return GitRun {
                code: -1,
                stdout: String::new(),
                stderr: e.to_string(),
            };
        }
    };
    GitRun {
        code: output.status.code().unwrap_or(-1),
        stdout: String::from_utf8_lossy(&output.stdout).to_string(),
        stderr: String::from_utf8_lossy(&output.stderr).to_string(),
    }
}

fn safe_relative(root: &str, rel: &str) -> Option<String> {
    let rel_path = Path::new(rel);
    if rel_path.is_absolute() {
        return None;
    }
    let abs: PathBuf = Path::new(root).join(rel_path);
    let canonical = abs.canonicalize().ok().unwrap_or(abs.clone());
    let root_canonical = Path::new(root).canonicalize().ok()?;
    if !canonical.starts_with(&root_canonical) {
        return None;
    }
    canonical
        .strip_prefix(&root_canonical)
        .ok()
        .map(|p| p.to_string_lossy().to_string())
}

fn parse_status(output: &str) -> GitStatusResult {
    let mut branch: Option<String> = None;
    let mut ahead = 0u32;
    let mut behind = 0u32;
    let mut files: Vec<GitFileStatus> = Vec::new();

    for line in output.split('\n') {
        if line.is_empty() {
            continue;
        }
        if let Some(rest) = line.strip_prefix("# branch.head ") {
            let name = rest.trim();
            branch = if name == "(detached)" {
                None
            } else {
                Some(name.to_string())
            };
        } else if let Some(rest) = line.strip_prefix("# branch.ab ") {
            // format: "+<ahead> -<behind>"
            for part in rest.split_whitespace() {
                if let Some(n) = part.strip_prefix('+') {
                    ahead = n.parse().unwrap_or(0);
                } else if let Some(n) = part.strip_prefix('-') {
                    behind = n.parse().unwrap_or(0);
                }
            }
        } else if line.starts_with("1 ") || line.starts_with("2 ") {
            let parts: Vec<&str> = line.split(' ').collect();
            let xy = parts.get(1).copied().unwrap_or("..");
            let mut chars = xy.chars();
            let x = chars.next().unwrap_or('.');
            let y = chars.next().unwrap_or('.');
            let path = if line.starts_with("2 ") {
                parts.get(parts.len().saturating_sub(2)).copied().unwrap_or("")
            } else {
                parts.last().copied().unwrap_or("")
            };
            if path.is_empty() {
                continue;
            }
            files.push(GitFileStatus {
                path: path.to_string(),
                index: x.to_string(),
                working_tree: y.to_string(),
                staged: x != '.' && x != '?',
            });
        } else if let Some(rest) = line.strip_prefix("? ") {
            files.push(GitFileStatus {
                path: rest.to_string(),
                index: "?".to_string(),
                working_tree: "?".to_string(),
                staged: false,
            });
        }
    }

    GitStatusResult {
        ok: true,
        branch,
        ahead,
        behind,
        files,
        error: None,
    }
}

/* ────────────────────────────────────────── *
 * Tauri commands                             *
 * ────────────────────────────────────────── */

#[tauri::command]
pub async fn scm_status(root_path: String) -> Result<GitStatusResult, String> {
    let rootPath = root_path;
    if rootPath.is_empty() {
        return Ok(GitStatusResult {
            ok: false,
            branch: None,
            ahead: 0,
            behind: 0,
            files: Vec::new(),
            error: Some("No workspace open".to_string()),
        });
    }
    let inside = run_git(
        &rootPath,
        &["rev-parse", "--is-inside-work-tree"],
        None,
        Duration::from_secs(3),
    );
    if inside.code != 0 {
        return Ok(GitStatusResult {
            ok: false,
            branch: None,
            ahead: 0,
            behind: 0,
            files: Vec::new(),
            error: Some("Not a git repository".to_string()),
        });
    }
    let status = run_git(
        &rootPath,
        &[
            "status",
            "--porcelain=v2",
            "--branch",
            "--untracked-files=all",
        ],
        None,
        Duration::from_secs(5),
    );
    if status.code != 0 {
        return Ok(GitStatusResult {
            ok: false,
            branch: None,
            ahead: 0,
            behind: 0,
            files: Vec::new(),
            error: Some(if status.stderr.trim().is_empty() {
                "git status failed".to_string()
            } else {
                status.stderr.trim().to_string()
            }),
        });
    }
    Ok(parse_status(&status.stdout))
}

#[tauri::command]
pub async fn scm_diff(
    root_path: String,
    path: String,
    staged: Option<bool>,
) -> Result<GitDiffResult, String> {
    let rootPath = root_path;
    let staged = staged.unwrap_or(false);
    let safe = match safe_relative(&rootPath, &path) {
        Some(s) => s,
        None => {
            return Ok(GitDiffResult {
                ok: false,
                diff: String::new(),
                error: Some("Path escapes workspace".to_string()),
            });
        }
    };
    let mut args: Vec<&str> = vec!["diff", "-U3", "--no-color"];
    if staged {
        args.push("--cached");
    }
    args.push("--");
    args.push(&safe);
    let r = run_git(&rootPath, &args, None, Duration::from_secs(10));
    // git diff returns 1 when there are differences (not an error).
    if r.code != 0 && r.code != 1 {
        return Ok(GitDiffResult {
            ok: false,
            diff: String::new(),
            error: Some(if r.stderr.trim().is_empty() {
                "git diff failed".to_string()
            } else {
                r.stderr.trim().to_string()
            }),
        });
    }
    Ok(GitDiffResult {
        ok: true,
        diff: r.stdout,
        error: None,
    })
}

#[tauri::command]
pub async fn scm_stage(
    root_path: String,
    args: StageArgs,
) -> Result<StageResult, String> {
    let rootPath = root_path;
    let mut paths: Vec<String> = Vec::new();
    for p in &args.paths {
        match safe_relative(&rootPath, p) {
            Some(s) => paths.push(s),
            None => {
                return Ok(StageResult {
                    ok: false,
                    error: Some(format!("Path escapes workspace: {}", p)),
                });
            }
        }
    }
    let mut argv: Vec<&str> = vec![if args.stage { "add" } else { "reset" }, "--"];
    for p in &paths {
        argv.push(p);
    }
    let r = run_git(&rootPath, &argv, None, Duration::from_secs(10));
    if r.code == 0 {
        Ok(StageResult { ok: true, error: None })
    } else {
        Ok(StageResult {
            ok: false,
            error: Some(r.stderr.trim().to_string()),
        })
    }
}

#[tauri::command]
pub async fn scm_commit(
    root_path: String,
    args: CommitArgs,
) -> Result<CommitResult, String> {
    let rootPath = root_path;
    let message = args.message.trim().to_string();
    if message.is_empty() {
        return Ok(CommitResult {
            ok: false,
            sha: None,
            error: Some("Commit message is empty".to_string()),
        });
    }
    if args.stage_all {
        let r = run_git(&rootPath, &["add", "-A"], None, Duration::from_secs(10));
        if r.code != 0 {
            return Ok(CommitResult {
                ok: false,
                sha: None,
                error: Some(if r.stderr.trim().is_empty() {
                    "git add -A failed".to_string()
                } else {
                    r.stderr.trim().to_string()
                }),
            });
        }
    }
    let r = run_git(
        &rootPath,
        &["commit", "-F", "-"],
        Some(&message),
        Duration::from_secs(15),
    );
    if r.code != 0 {
        let err = if !r.stderr.trim().is_empty() {
            r.stderr.trim().to_string()
        } else if !r.stdout.trim().is_empty() {
            r.stdout.trim().to_string()
        } else {
            "git commit failed".to_string()
        };
        return Ok(CommitResult {
            ok: false,
            sha: None,
            error: Some(err),
        });
    }
    let sha = run_git(&rootPath, &["rev-parse", "HEAD"], None, Duration::from_secs(3));
    Ok(CommitResult {
        ok: true,
        sha: if sha.code == 0 {
            Some(sha.stdout.trim().to_string())
        } else {
            None
        },
        error: None,
    })
}

#[tauri::command]
pub async fn scm_run(root_path: String, cmd: String) -> Result<RunResult, String> {
    let rootPath = root_path;
    let allowed = matches!(cmd.as_str(), "pull" | "push" | "fetch");
    if !allowed {
        return Ok(RunResult {
            ok: false,
            stdout: None,
            stderr: None,
            error: Some(format!("Command not allowed: {}", cmd)),
        });
    }
    let r = run_git(&rootPath, &[cmd.as_str()], None, Duration::from_secs(60));
    Ok(RunResult {
        ok: r.code == 0,
        stdout: Some(r.stdout.clone()),
        stderr: Some(r.stderr.clone()),
        error: if r.code == 0 {
            None
        } else if r.stderr.trim().is_empty() {
            Some(format!("git {} failed", cmd))
        } else {
            Some(r.stderr.trim().to_string())
        },
    })
}
